product/tests/test_product.py

```diff
@@ -16,20 +17,34 @@ def product0():
     return Product(qr_code=QR, product_name=NAME, description=DESCRIPTION)
 
 
+@pytest.fixture
+def product1():
+    return Product(qr_code=QR1, product_name=NAME, description=DESCRIPTION)
+
+
 @pytest.fixture
 def saved_product0(product0):
     product0.save_product()
     return product0
 
 
+@pytest.fixture
+def saved_product1(product1):
+    product1.save_product()
+    return product1
+
+
 class TestProductModel:
     @pytest.mark.django_db()
-    def test_save_product_and_delete_product(self, saved_product0):
+    def test_save_product_and_delete_product(self, saved_product0, saved_product1):
         assert saved_product0 in Product.objects.all()
-        Product.delete_product(saved_product0.qr_code)
+        saved_product0.delete()
         assert saved_product0 not in Product.objects.all()
-        saved_product0.save_product()
-        assert saved_product0 in Product.objects.all()
+        with pytest.raises(Product.DoesNotExist):
+            test_doesnotexist = Product("DOESNOTEXIST", "TEST", "TEST")
+            test_doesnotexist.delete()
+            raise Product.DoesNotExist()
```

Comment:
The code should do the raising rather than the test.

Reply:
I dont understand,can you explain?

Reply:
We use that context manager (`with pytest.raises(SomeExceptionClass)`) when we test a scenario in which the code that we test should fail and raised an exception of that class. The context manager asserts that the exception was raised.

If you raise the exception inside the test, then your test will always pass. It's the equivalent of writing `assert True` on your test.

Does the built-in `delete` method raises something when deleting a non-existing product? If not, then you don't need that context manager.

Reply:
The code first creates an instance of the Product class called test_doesnotexist with the name "DOESNOTEXIST", and then calls the delete() method on that instance. If the delete() method raises a Product.DoesNotExist exception, the code will continue to run. If the exception is not raised, the pytest library will raise a Failed error

Reply:
How is that helping us? that last line makes the test to always pass. You could put any code between the `with` statement and the `raise` statement, and the test will pass.

Reply:
Sorry,buy i really cant understand what is wrong and why it is always passing as you are saying

Reply:
The idea of having that context manager is for testing that the exception was indeed raised. I can delete the 2 lines calling the code being tested and the test will still pass, so what does it test?
```
        with pytest.raises(Product.DoesNotExist):
            raise Product.DoesNotExist()
```

Reply:
try:
            test_doesnotexist = Product("DOESNOTEXIST", "TEST", "TEST")
            test_doesnotexist.delete_product()
        except Product.DoesNotExist:
            assert False
    def delete_product(self):
        Product.objects.get(qr_code=self.qr_code).delete()

this code raise Product Does Not Exist,i have converted that to context manager and it passes

Reply:
regular delete does not raise the does not exist as i tried,if you want it to be raised first you need to get the row with the specific qr code and then delete it if it already exist

Reply:
When is the `DoesNotExist` exception being raised? When you run the test, you should know whether it need to be raised or not, and assert accordingly.

Reply:
In the new code the doesnotexist is being raised that is why the test have completed,other wise the test should fail i dont need to assert nothing i guess,since the product does not exist raised assert true,other wise assert false or exception

Reply:
I lost you.

Do you want your code to raise an exception when you delete? Are you currently just adding a new method `delete_product` so it will raise an exception?

I honestly don't know if you should have it raise an exception or not. It's up to you to decide. I'd say:
- If you know that you want it to raise an exception, then fine.
- If you don't know that you need that, then don't include that code (you could add it later on if needed).
- If you don't add the code that raises an exception, then don't test that the an exception is being raised.

Reply:
- I need that code and yes i wanted that exception to be raised and the exception is tested we can see the test passing
- I dont want a situation that you are ready to delete a non existed prouduct

Reply:
ok. I'm fine with that I think. I'm going to add some comments on the current code (sorry about that).